web/src/analytics.ts

```diff
@@ -0,0 +1,42 @@
+import { useConnectionState } from "@/data/hooks";
+import { userSessionID } from "@/data/recoil";
+import { useEffect } from "react";
+import { useLocation } from "react-router-dom";
+import { useRecoilState } from "recoil";
+import { v4 as uuidv4 } from "uuid";
+
+export function trackAnalyticsEvent(
+  event: string,
+  params?: Record<string, string | number>
+) {
+  window.analytics.track(`rtdm-${event}`, params);
+}
+
+export function useAnalytics() {
+  const location = useLocation();
+  const { connectionType } = useConnectionState();
+  const [userID, setUserID] = useRecoilState(userSessionID);
+
+  useEffect(() => {
+    const { pathname } = location;
+    trackAnalyticsEvent("change-page", { pathname });
+  }, [location]);
+
+  useEffect(() => {
+    if (connectionType) {
+      trackAnalyticsEvent("connection-successful", { connectionType });
+    }
+  }, [connectionType]);
+
+  useEffect(() => {
+    if (userID === "") {
+      const newUserID = uuidv4();
+      setUserID(newUserID);
+    }
+    if (window.analytics) {
+      window.analytics.identify(userID);
+    } else {
+      console.warn("Analytics are not defined");
+    }
+  }, [userID, setUserID]);
```

Comment:
so as it's currently written you will cause identify to fire twice on first load - once with an undefined userID and once with it defined

a better solution is to just use `uuidv4()` as the default value for the userID atom. it's a small additional cost per page load for simpler code :)

Reply:
it seems that segment is smart enough to not log empty user IDs. Either way, I think it's more clean to have it as default value as you suggest. Changed it!